1.5/Source/PawnStorages/PawnStorages/Farm/ITab_Farm.cs

```diff
@@ -1,119 +1,130 @@
-﻿using System.Collections.Generic;
-using System.Text;
+﻿using System;
+using System.Collections.Generic;
+using System.Data.SqlTypes;
+using System.Linq;
+using System.Security.Cryptography;
+using PawnStorages.Farm.Comps;
+using PawnStorages.Farm.Interfaces;
 using RimWorld;
+using TMPro;
 using UnityEngine;
 using Verse;
-using Color = UnityEngine.Color;
+using Verse.Noise;
+using Verse.Sound;
+using static HarmonyLib.Code;
+using static UnityEngine.Random;
 
 namespace PawnStorages.Farm
 {
     public class ITab_Farm : ITab
     {
-        private static readonly Vector2 WinSize = new(300f, 480f);
-        public readonly ThingFilterUI.UIState ThingFilterState = new();
-        public const float lineHeight = 60f;
-        private bool alternate;
-
-        public CompFarmStorage compFarmStorage => SelThing.TryGetComp<CompFarmStorage>();
-        public CompFarmNutrition compFarmNutrition => SelThing.TryGetComp<CompFarmNutrition>();
-        public bool NeedsDrop => PawnStoragesMod.settings.AllowNeedsDrop || compFarmStorage.Props.needsDrop;
-
         public ITab_Farm()
         {
             size = ITab_PenAnimals.WinSize;
-            labelKey = "TabPenAnimals";
+            labelKey = "PS_FarmTab";
         }
 
-        public float PawnFullness(Pawn pawn)
-        {
-            if (pawn.TryGetComp(out CompEggLayer compLayer))
-            {
-                return compLayer.eggProgress;
-            }
+        public IFarmTabParent Parent => SelThing as IFarmTabParent;
 
-            if (pawn.TryGetComp(out CompHasGatherableBodyResource compGatherable))
-            {
-                return compGatherable.Fullness;
-            }
+        protected float TopAreaHeight => 20f;
 
-            return 0;
-        }
+        protected const float LineHeight = 20f;
+        protected Vector2 ScrollPosition = Vector2.zero;
+        protected QuickSearchWidget QuickSearchWidget = new QuickSearchWidget();
 
-        public bool DrawLine(float position, float width, Pawn pawn)
+        public override void FillTab()
         {
-            Rect rect = new Rect(0.0f, position, width, lineHeight);
 
-            if (alternate)
+            Rect tabRect = new Rect(0.0f, 0.0f, ITab_Storage.WinSize.x, ITab_Storage.WinSize.y).ContractedBy(10f);
+            Widgets.BeginGroup(tabRect);
+
+            Rect menuRect = new Rect(0.0f, 20f, tabRect.width, tabRect.height - 20f);
+            Widgets.DrawMenuSection(menuRect);
+            float num1 = menuRect.width - 2f;
+            Rect buttonRect = new Rect(menuRect.x + 3f, menuRect.y + 3f, (float)((double)num1 / 2.0 - 3.0 - 1.5), 24f);
+            if (Widgets.ButtonText(buttonRect, (string)"ClearAll".Translate()))
             {
-                Widgets.DrawRectFast(rect, new Color(1f, 1f, 1f, ITab_Pawn_Log_Utility.AlternateAlpha));
+                // filter.SetDisallowAll(forceHiddenDefs, forceHiddenFilters);
```

Comment:
these buttons seem to do nothing? Just play the sound?

Reply:
Fixed. Added the code to clear/add all, forgot to call it :)